scipy/sparse/linalg/_eigen/lobpcg/lobpcg.py

```diff
@@ -67,20 +66,38 @@ def _makeMatMat(m):
         return lambda v: m @ v
 
 
+def _dot_inplace(x, y, verbosityLevel=0):
+    """Perform 'np.dot' in-place if dtypes match."""
+    try:
+        np.dot(x, y, out=x)
+    except Exception:
+        if verbosityLevel:
+            warnings.warn(
+                f"Trying in-place x = x @ y, possibly their shapes "
+                f"{x.shape} and {y.shape} mismatch, or x.dtype {x.dtype} "
+                f"does not match y.dtype {y.dtype}, "
+                f"or x is a slice rather than a C-Array, "
+                f"so x needs to be overwritten preventing in-place.",
+                UserWarning, stacklevel=3
+            )
+        x = x @ y
```

Comment:
So generally, `except Exception:` is viewed as an anti-pattern because it's way too broad. In this particular instance, I feel this is even more the case, because we know a-priori which conditions are going to make the inplace-operation fail.

This should be even faster, because some trivial equality checks should be quicker than calling a function, failing, and then handling that.

Why not do
```suggestion
    if x.shape == y.shape and _dtype_compatible(x.dtype, y.dtype) and _is_not_a_slice(x):
        np.dot(x, y, out=x)
    else:
        x = x @ y
```

If we really really want, we could still have the try-except logic in the else block.

Reply:
@h-vetinari Thanks, I can try this. I am unsure if this is an exhaustive list, however. I got this by just trial-and-error. 

I would still like to keep the warning though for the user to known that the code may run out of memory because the operation is not in-place, if it's ok?

Reply:
Well, the more conditions you can identify in advance, the more should move into explicit checks. In an ideal world, no try/except would be necessary in the else branch, but it's fine IMO to keep the fall-back.

Reply:
@h-vetinari :
-  I am unfamiliar with `_dtype_compatible` and unsure if this condition is sufficient for the in-place dot to run. In-place evidently requires `x.dtype` to be the same as `(x @ y).dtype`
- your condition `x.shape == y.shape` looks strange for the x @ y matrix product, in-place or not. Should not it be `y.shape[0] == y.shape[1]` instead?

Reply:
Sorry, I should have been more clear, both `_dtype_compatible` and `_is_not_a_slice` were just dummies for exposition. I'm pretty sure that numpy would expose enough functionality and information that these functions shouldn't be very complicated.

You're right about the matrix-matrix multiplication constraints, though for `X` an MxN matrix to be updated inplace, I assume both M & N need to stay the same?

Reply:
You proposed changes in a format of a commit suggestion, so I took it literally. I do not how to actually check what you mean by `_dtype_compatible` and `_is_not_a_slice`

For `x`, an MxN matrix, to be updated in-place by the product `x @ y`, it is the product that must be the same MxN shape as x, which implies that `y` must be NxN according to the rules of the product of matrices. I have implemented this change in the latest edit, and it results in no errors - thanks for your comment!

Reply:
> I do not how to actually check what you mean by `_dtype_compatible` and `_is_not_a_slice`

OK, I think we can do two things here. `_dtype_compatible` was meant to capture the fact that probably dtype _equality_ is not _always_ necessary to enable inplace updates. However, since we keep the try/except fallback, we can check for dtype-equality in the fast-path, and leave all other cases to the fallback. WDYT?